Add businessDiff method

kristremblay opened the pull request:
Added businessDiff method to get number of business days between DateTimes.

After completing the switch from Moment to Luxon, I realized this functionality was missing in this library.  The solution was to borrow from https://github.com/kalmecak/moment-business-days/blob/master/index.js.

Usage:

```js
const targetDate = DateTime.local().plusBusiness({ days: 10 });

const businessDaysDiff = DateTime.local().businessDiff(targetDate);

// businessDaysDiff = 10
```

kristremblay committed:
Added businessDiff method to get number of buiness days
between DateTimes.

kristremblay committed:
Fixed failing test due to time issue.
In practice this would be handled based on developer needs.

kristremblay committed:
Clean tests, fixed bug in while to incl. last day.

kristremblay commented:
@amaidah oddly, the tests were passing when this was submitted but when I opened vscode today they were failing.  I've refactored to clear out some of the spider webs left over from a day of working :)

kristremblay committed:
Reverted < to <= in day count loop.

amaidah commented:
Hey, thanks for opening a PR -- I will take a look (can't promise this weekend). I remember thinking about this method originally and had thought about changing the interface a bit.

kristremblay commented:
Sounds good, and you're welcome 😄

kristremblay commented:
Strangely, if the initial date is not a business day, the calculation ends up coming up one short.

Reproduce:

```js
const diff = DateTime.local(2021, 2, 28).businessDiff(DateTime.local(2021, 3, 2));
// expected: 2
// received: 1
```

kristremblay committed:
Replaced relative param with config object
added includeEndDate

kristremblay commented:
Swapped out the relative param for a config object.

kristremblay committed:
Fix break in documentation.

kristremblay committed:
Update businessDiff description.

kristremblay committed:
Improved BusinessDiffConfig description

kristremblay commented:
@amaidah sorry to bother, but have you had a chance to take a peek?

amaidah commented:
> @amaidah sorry to bother, but have you had a chance to take a peek?

Thanks for being super patient, not yet. Will try for this week

amaidah commented on src/index.test.js:
```diff
@@ -324,3 +324,59 @@ describe('time zone is carried over after a business-day operation', () => {
   });
 });
+
+describe('business day diff', () => {
```
```suggestion
describe('diffBusiness()', () => {
```

amaidah commented on src/index.js:
```diff
@@ -146,4 +146,48 @@ DateTime.prototype.minusBusiness = function({ days = ONE_DAY } = {}) {
+ * @param {BusinessDiffConfig} config
+ * @returns {number}
+ */
+DateTime.prototype.businessDiff = function(
```
```suggestion
DateTime.prototype.diffBusiness = function(
```

amaidah commented on src/index.js:
```diff
@@ -146,4 +146,48 @@ DateTime.prototype.minusBusiness = function({ days = ONE_DAY } = {}) {
+/**
+ * @typedef BusinessDiffConfig
+ * @property {boolean} [includeEndDate=false] - include the end date in the calculation
+ * @property {boolean} [relative=false] - signs the return value as negative if end date is in the past
```
If I think about how I use a calendar, I say things like 2 days ago (-2) so it doesn't make total sense to me that a difference would be absolute by default. Is there a reason for this? I remember being confused with the moment version of this plugin when I first saw that too.

Both [diff](https://moment.github.io/luxon/docs/class/src/datetime.js~DateTime.html#instance-method-diff) and [diffNow](https://moment.github.io/luxon/docs/class/src/datetime.js~DateTime.html#instance-method-diffNow) are relative and I'd like to to keep this plugin similar to Luxon core concepts

kristremblay replied:
I think the rationale is that you already have methods for determining whether a date is in the past and it is _probably_ a less common scenario to need a negative value for a business day calculation.

That being said, consistency is a really strong argument so if you think it would be better defaulting relative to true I will support that.

amaidah replied:
Ok cool, let's go with relative as default

amaidah commented on src/index.js:
```diff
@@ -146,4 +146,48 @@ DateTime.prototype.minusBusiness = function({ days = ONE_DAY } = {}) {
 
+/**
+ * @typedef BusinessDiffConfig
+ * @property {boolean} [includeEndDate=false] - include the end date in the calculation
```
This is a neat concept I didn't see as an option in moment. What kind of scenarios would require including an end date?

kristremblay replied:
I've had business day requirements in the past where the business wanted to include the end date.  For example, if you have a report and you are entering how long some job is going to take in business days, including end date allows you to say that a task will be completed by end of day on the target business day.  That would make a job targeted for Friday have a time value of 2 business days if entered on the Thursday.

I'm sure there are other applications, but this was the one that immediately came to mind, so hopefully that makes sense :)

amaidah replied:
ok makes sense

amaidah commented on src/index.test.js:
```diff
@@ -324,3 +324,59 @@ describe('time zone is carried over after a business-day operation', () => {
+
+  it('knows there are 3 business days between two dates that are 3 business days apart', () => {
+    const myCompanyTakesNoHolidays = [];
+    const startDate = DateTime.local().startOf('day');
```
is `.startOf` necessary for the tests?

kristremblay replied:
Removing them on local seems to not affect the tests, perhaps we can refactor them out.

kristremblay committed:
Update src/index.test.js

Co-authored-by: Andrew Maidah <a.maidah@gmail.com>

kristremblay committed:
Update src/index.js

Co-authored-by: Andrew Maidah <a.maidah@gmail.com>

kristremblay committed:
Rename businessDiff  diffBusiness for consistency

kristremblay committed:
Readme regen

kristremblay committed:
Removed start of day from tests

amaidah commented on src/index.test.js:
```diff
@@ -324,3 +324,57 @@ describe('time zone is carried over after a business-day operation', () => {
+
+describe('diffBusiness()', () => {
+  it('knows two identical DateTimes have a business day diff of 0', () => {
+    const targetDate = DateTime.local().plus({ hours: 2 });
```
Actually, this one with the hours might actually still need `.startOf` because the test could 2 hours into the next day. The other tests that add days should be okay.

kristremblay replied:
Agreed

amaidah commented on src/index.test.js:
```diff
@@ -324,3 +324,57 @@ describe('time zone is carried over after a business-day operation', () => {
+
+describe('diffBusiness()', () => {
+  it('knows two identical DateTimes have a business day diff of 0', () => {
+    const targetDate = DateTime.local().plus({ hours: 2 });
```
```suggestion
    const targetDate = DateTime.local().startOf('day').plus({ hours: 2 });
```

kristremblay replied:
I committed your suggestion, feel free to resolve at your convenience.

kristremblay committed:
Update src/index.test.js

Put startOf call back in.

Co-authored-by: Andrew Maidah <a.maidah@gmail.com>

kristremblay commented:
Hmm...running into some failing tests when the current date is not a valid business day.  Going to investigate.

kristremblay commented:
Ok upon further investigation, it was already working correctly, it was just a matter of a bad test.

I am interested in your input on this, but maybe we should have a separate test for starting on a non-business day.

amaidah commented:
> How should calculations work if performed on a non-business day?

Should just return the number of days that are business days between the two compared dates right? So if starting on Sat, and the next day is Sun and both are nonbusiness running a `diffBusiness` on that range should return 0. I'm curious what would happen though if `includeEndDate` were true in this scenario

> I am interested in your input on this, but maybe we should have a separate test for starting on a non-business day.

Yes, now that you bring it up, sounds smart to test for a few more things:

* Starts on a non-business day (in general)
* Ends on non-business day (in general)
* Starts on business day but ends on non-business day and includes end date
* Start on non-business day and ends on non-business day with no business days in the range and includes end date


I also include error handling conditional for the other methods:

```javascript
  if (!dt.isValid) {
    return dt;
  }
```

Maybe we should think about returning an error if trying to call `diffBusiness` on non something that isn't a DateTime

kristremblay commented:
> I'm curious what would happen though if includeEndDate were true in this scenario

That's actually a good point that isn't currently accounted for.  I just added the following:

```js
let daysDiff = Number(
  includeEndDate && end.isBusinessDay() && !end.isHoliday()
);
```

Testing for starting on a non-business day should be next on the list.  I will work on that over the weekend.  In the meantime, I've updated the branch to use a static start date of Monday May 3 2021.

kristremblay committed:
Test cleanup and includeEndDate fix

kristremblay committed:
Fixed missing coverage

amaidah commented:
> > I'm curious what would happen though if includeEndDate were true in this scenario
> 
> That's actually a good point that isn't currently accounted for. I just added the following:
> 
> ```js
> let daysDiff = Number(
>   includeEndDate && end.isBusinessDay() && !end.isHoliday()
> );
> ```
> 
> Testing for starting on a non-business day should be next on the list. I will work on that over the weekend. In the meantime, I've updated the branch to use a static start date of Monday May 3 2021.

FYI, I merged in a bunch of dependabot PRs so you may need to pull from master

kristremblay commented:
Thanks for letting me know!

hdoan-codaio commented:
I'm interested in using this method. Can you merge the PR?
